validator/sawtooth_validator/execution/executor.py

```diff
@@ -304,37 +318,71 @@ def _execute_schedule(self):
 
             # Since we have already checked if the transaction should be failed
             # all other cases should either be executed or waited for.
-            self._execute_or_wait_for_processor_type(
-                processor_type=processor_type,
-                content=content,
-                signature=txn.header_signature)
+            if processor_type in self._processors:
+                self._execute(
+                    processor_type=processor_type,
+                    content=content,
+                    signature=txn.header_signature)
+            else:
+                self._wait_for_processor_registration(
+                    processor_type=processor_type,
+                    content=content,
+                    signature=txn.header_signature)
 
         self._done = True
 
-    def _execute_or_wait_for_processor_type(
-            self, processor_type, content, signature):
-        processor = self._processors.get_next_of_type(
-            processor_type=processor_type)
-        if processor is None:
-            LOGGER.debug("no transaction processors registered for "
-                         "processor type %s", processor_type)
-            if processor_type not in self._waiters_by_type:
-                in_queue = queue.Queue()
-                in_queue.put_nowait((content, signature))
-                waiter = _Waiter(
-                    self._send_and_process_result,
-                    processor_type=processor_type,
-                    processors=self._processors,
-                    in_queue=in_queue,
-                    waiters_by_type=self._waiters_by_type)
-                self._waiters_by_type[processor_type] = waiter
-                self._waiting_threadpool.submit(waiter.run_in_threadpool)
-            else:
-                self._waiters_by_type[processor_type].add_to_in_queue(
-                    (content, signature))
-        else:
+    def _execute(self, processor_type, content, signature):
+        try:
+            processor = self._processors.get_next_of_type(
+                processor_type=processor_type)
             connection_id = processor.connection_id
             self._send_and_process_result(content, connection_id, signature)
+        except ProcessorQueuesFullError:
+            self._wait_for_processor_vacancy(
+                processor_type=processor_type,
+                content=content,
+                signature=signature)
+
+    def _wait_for_processor_vacancy(self, processor_type, content, signature):
+        LOGGER.debug("all transaction processors of type "
+                     "%s are at max occupancy", processor_type)
+        if processor_type not in self._waiters_by_type_vacancy:
```

Comment:
I think this is a data race, since if there are multiple threads executing it multiple threads could see there is no waiter and create them and then overwrite the others when they are set in the `_waiters_by_type_vacancy`. It looks like this was an existing data race.

Reply:
Since this was an existing data race, I will accept this if you open a bug report on JIRA.